### PR TITLE
BUG: Cannot sample on DataFrameGroupBy with weights when index is specified

### LINKED ISSUE #39927: BUG: Cannot sample on DataFrameGroupBy with weights when index is specified
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
import pandas as pd
df1 = pd.DataFrame({'a': [3, 5, 2, 7, 1, 2, 4, 6, 2, 4],
                    'b': [8, 3, 4, 1, 8, 2, 2, 2, 2, 3],
                    'c': [1, 8, 8, 5, 3, 5, 0, 0, 5, 4],
                    'd': [9, 7, 6, 3, 6, 3, 2, 1, 5, 8]}, index=list(range(10, 20)))
df1.groupby('c').sample(1, weights=df1['d'].to_numpy())
```

#### Problem description

Sampling ought to work but exception is reported with error stack below:

```
Traceback (most recent call last):
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3417, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-4-1fc31a504740>", line 1, in <module>
    df1.groupby('c').sample(1, weights=df1['d'].to_numpy())
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/groupby/groupby.py", line 3024, in sample
    ws = [weights[idx] for idx in self.indices.values()]
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/groupby/groupby.py", line 3024, in <listcomp>
    ws = [weights[idx] for idx in self.indices.values()]
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/series.py", line 875, in __getitem__
    return self._get_with(key)
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/series.py", line 910, in _get_with
    return self.loc[key]
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/indexing.py", line 895, in __getitem__
    return self._getitem_axis(maybe_callable, axis=axis)
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/indexing.py", line 1113, in _getitem_axis
    return self._getitem_iterable(key, axis=axis)
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/indexing.py", line 1053, in _getitem_iterable
    keyarr, indexer = self._get_listlike_indexer(key, axis, raise_missing=False)
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/indexing.py", line 1266, in _get_listlike_indexer
    self._validate_read_indexer(keyarr, indexer, axis, raise_missing=raise_missing)
  File "/Users/wenjun/miniconda3/lib/python3.8/site-packages/pandas/core/indexing.py", line 1308, in _validate_read_indexer
    raise KeyError(f"None of [{key}] are in the [{axis_name}]")
KeyError: "None of [Int64Index([6, 7], dtype='int64')] are in the [index]"
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 7d32926db8f7541c356066dcadabf854487738de
python           : 3.8.5.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.6.0
Version          : Darwin Kernel Version 19.6.0: Tue Jan 12 22:13:05 PST 2021; root:xnu-6153.141.16~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : zh_CN.UTF-8
pandas           : 1.2.2
numpy            : 1.19.1
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 20.2.2
setuptools       : 49.6.0.post20200814
Cython           : 0.29.21
pytest           : 6.0.1
hypothesis       : None
sphinx           : 3.2.1
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.17.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : 0.4.2
gcsfs            : None
matplotlib       : 3.2.2
numexpr          : 2.7.1
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 1.0.1
pyxlsb           : None
s3fs             : None
scipy            : 1.6.0
sqlalchemy       : 1.3.18
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : 0.51.2

</details>

### PR DESCRIPTION
**Pull request by wjsi**
- [x] closes #39927
- [x] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/groupby/test_sample.py [hunk 1]
```diff
@@ -116,14 +116,18 @@ def test_groupby_sample_without_n_or_frac():
     tm.assert_series_equal(result, expected)
 
 
-def test_groupby_sample_with_weights():
+@pytest.mark.parametrize(
+    "index, expect_index",
+    [(["w", "x", "y", "z"], ["w", "w", "y", "y"]), ([3, 4, 5, 6], [3, 3, 5, 5])],
+)
+def test_groupby_sample_with_weights(index, expect_index):
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_sample.py:123]: Can you add issue number in a comment as the first line of this function (i.e. `GH ##39927`)

File: pandas/tests/groupby/test_sample.py [hunk 2]
```diff
@@ -116,14 +116,18 @@ def test_groupby_sample_without_n_or_frac():
     tm.assert_series_equal(result, expected)
 
 
-def test_groupby_sample_with_weights():
+@pytest.mark.parametrize(
+    "index, expect_index",
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_sample.py:120]: nit: Can you change to expected_index

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the PR! Changes look really good, two minor requests.

**Comment by rhshadrach** [comment]
Also need a note in the 1.3.0 whatsnew under GroupBy Bug Fixes

**Review by rhshadrach — APPROVED** [review]
lgtm - cc @jreback

**Review by jreback — APPROVED** [review]
lgtm. merge on green.

**Comment by rhshadrach** [comment]
Thanks @wjsi!

**Pull request merged**